# Analyzing Tabular Data using Python and Pandas

![](https://i.imgur.com/zfxLzEv.png)

This tutorial covers the following topics:

- Reading a CSV file into a Pandas data frame
- Retrieving data from Pandas data frames
- Querying, sorting, and analyzing data
- Merging, grouping, and aggregation of data
- Extracting useful information from dates
- Basic plotting using line and bar charts
- Writing data frames to CSV files

## Reading a CSV file using Pandas

[Pandas](https://pandas.pydata.org/) is a popular Python library used for working in tabular data (similar to the data stored in a spreadsheet). Pandas provides helper functions to read data from various file formats like CSV, Excel spreadsheets, HTML tables, JSON, SQL, and more. Let's download a file `italy-covid-daywise.txt` which contains day-wise Covid-19 data for Italy in the following format:

```
date,new_cases,new_deaths,new_tests
2020-04-21,2256.0,454.0,28095.0
2020-04-22,2729.0,534.0,44248.0
2020-04-23,3370.0,437.0,37083.0
2020-04-24,2646.0,464.0,95273.0
2020-04-25,3021.0,420.0,38676.0
2020-04-26,2357.0,415.0,24113.0
2020-04-27,2324.0,260.0,26678.0
2020-04-28,1739.0,333.0,37554.0
...
```

This format of storing data is known as *comma-separated values* or CSV. 

> **CSVs**: A comma-separated values (CSV) file is a delimited text file that uses a comma to separate values. Each line of the file is a data record. Each record consists of one or more fields, separated by commas. A CSV file typically stores tabular data (numbers and text) in plain text, in which case each line will have the same number of fields. (Wikipedia)

To read the file, we can use the `read_csv` method from Pandas. First, let's import the Pandas library. As a convention, it is imported with the alias `pd`.

#### TODO

import pandas

In [2]:
import pandas as pd

#### TODO:

read the italy_covid_daywise csv file into a dataframe.

Call the dataframe covid_df

In [3]:
covid_df = pd.read_csv('italy-covid-daywise.csv')
#Can specify names explicity with: name['name_header', 'nextnameheader']

Data from the file is read and stored in a `DataFrame` object - one of the core data structures in Pandas for storing and working with tabular data. We typically use the `_df` suffix in the variable names for dataframes.

In [4]:
covid_df

date  new_cases  new_deaths  new_tests
0    2020-02-22       14.0         0.0        NaN
1    2020-02-23       62.0         2.0        NaN
2    2020-02-24       53.0         0.0        NaN
3    2020-02-25       97.0         4.0        NaN
4    2020-02-26       93.0         5.0        NaN
..          ...        ...         ...        ...
190  2020-08-30     1444.0         1.0    53541.0
191  2020-08-31     1365.0         4.0    42583.0
192  2020-09-01      996.0         6.0    54395.0
193  2020-09-02      975.0         8.0        NaN
194  2020-09-03     1326.0         6.0        NaN

[195 rows x 4 columns]

#### TODO:

What is the type of covid_df

In [5]:
type(covid_df)

pandas.core.frame.DataFrame

#### TODO:

Print the head() of covid_df

In [6]:
covid_df.head() #Displays the first 5 rows

date  new_cases  new_deaths  new_tests
0  2020-02-22       14.0         0.0        NaN
1  2020-02-23       62.0         2.0        NaN
2  2020-02-24       53.0         0.0        NaN
3  2020-02-25       97.0         4.0        NaN
4  2020-02-26       93.0         5.0        NaN

Here's what we can tell by looking at the dataframe:

- The file provides day-wise counts for COVID-19 in Italy
- The metrics reported are new cases, deaths, and tests

Keep in mind that these are officially reported numbers. The actual number of cases & deaths may be higher, as not all cases are diagnosed. 

We can view some basic information about the data frame using the `.info` method.

#### TODO:

Check the covid_df information using .info()

In [7]:
covid_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 195 entries, 0 to 194
Data columns (total 4 columns):
 #   Column      Non-Null Count  Dtype  
---  ------      --------------  -----  
 0   date        195 non-null    object 
 1   new_cases   195 non-null    float64
 2   new_deaths  195 non-null    float64
 3   new_tests   135 non-null    float64
dtypes: float64(3), object(1)
memory usage: 6.2+ KB


#### TODO:

Check the covid_df columns data types using .dtypes

In [8]:
covid_df.dtypes #Checks column data types

date           object
new_cases     float64
new_deaths    float64
new_tests     float64
dtype: object

<img src="pandas-dtypes.png" width=400 height=400 />

#### TODO:

select just object columns


In [9]:
covid_df.select_dtypes(include='object') #In this case only date is an object, can also exclude=
covid_df.select_dtypes(include='object').head() #Guves first 5 rows, can keep stacking functions on top of each other
covid_df.select_dtypes(include='object').head().iloc[0] #Gives first row

date    2020-02-22
Name: 0, dtype: object

#### TODO:

select the numeric columns


In [10]:
covid_df.select_dtypes(exclude='object').head() #Head gives first 5 rows, can also include='number'

new_cases  new_deaths  new_tests
0       14.0         0.0        NaN
1       62.0         2.0        NaN
2       53.0         0.0        NaN
3       97.0         4.0        NaN
4       93.0         5.0        NaN

It appears that each column contains values of a specific data type. You can view statistical information for numerical columns (mean, standard deviation, minimum/maximum values, and the number of non-empty values) using the `.describe` method.

#### TODO:

Use .describe() to get a description of the covid_df dataframe

In [11]:
covid_df.describe() #Descriptive statistics with .describe()

new_cases  new_deaths     new_tests
count   195.000000  195.000000    135.000000
mean   1392.369231  182.035897  31699.674074
std    1631.028060  241.969653  11622.209757
min    -148.000000  -31.000000   7841.000000
25%     253.000000    9.000000  25259.000000
50%     584.000000   53.000000  29545.000000
75%    2025.500000  271.500000  37711.000000
max    6557.000000  971.000000  95273.000000

The `columns` property contains the list of columns within the data frame.

#### TODO:

List all columns in covid_df

In [12]:
covid_df.columns

Index(['date', 'new_cases', 'new_deaths', 'new_tests'], dtype='object')

You can also retrieve the number of rows and columns in the data frame using the `.shape` method.

#### TODO:

Get the shape of covid_df

In [13]:
covid_df.shape #Number of rows, number of columns, use shape when you want to loop through data frame outer = 0, inner = 1

(195, 4)

Here's a summary of the functions & methods we've looked at so far:

* `pd.read_csv` - Read data from a CSV file into a Pandas `DataFrame` object
* `.info()` - View basic infomation about rows, columns & data types
* `.describe()` - View statistical information about numeric columns
* `.columns` - Get the list of column names
* `.shape` - Get the number of rows & columns as a tuple


## Retrieving data from a data frame

The first thing you might want to do is retrieve data from this data frame, e.g., the counts of a specific day or the list of values in a particular column. To do this, it might help to understand the internal representation of data in a data frame. Conceptually, you can think of a dataframe as a dictionary of lists: keys are column names, and values are lists/arrays containing data for the respective columns. 

In [14]:
# Pandas format is simliar to this
covid_data_dict = {
    'date':       ['2020-08-30', '2020-08-31', '2020-09-01', '2020-09-02', '2020-09-03'],
    'new_cases':  [1444, 1365, 996, 975, 1326],
    'new_deaths': [1, 4, 6, 8, 6],
    'new_tests': [53541, 42583, 54395, None, None]
}

Representing data in the above format has a few benefits:

* All values in a column typically have the same type of value, so it's more efficient to store them in a single array.
* Retrieving the values for a particular row simply requires extracting the elements at a given index from each column array.


With the dictionary of lists analogy in mind, you can now guess how to retrieve data from a data frame. For example, we can get a list of values from a specific column using the `[]` indexing notation.

#### TODO:

Access the 'new_cases' key in covid_data_dict

In [15]:
covid_data_dict

{'date': ['2020-08-30',
  '2020-08-31',
  '2020-09-01',
  '2020-09-02',
  '2020-09-03'],
 'new_cases': [1444, 1365, 996, 975, 1326],
 'new_deaths': [1, 4, 6, 8, 6],
 'new_tests': [53541, 42583, 54395, None, None]}

In [16]:
covid_data_dict['new_cases']

[1444, 1365, 996, 975, 1326]

In [17]:
covid_data_dict['new_cases'][0]

1444

In [18]:
covid_data_dict['new_cases'][-1]

1326

#### TODO:

Access the first element in the 'new_cases' key in covid_data_dict

In [19]:
type(covid_df.new_cases)

pandas.core.series.Series

Each column is represented using a data structure called `Series`, which is essentially a numpy array with some extra methods and properties.

#### TODO:

Get the type of the 'new_cases' column in the covid_df dataframe

In [20]:
type(covid_df.new_cases)

pandas.core.series.Series

#### TODO:

Get the type of the covid_df

In [21]:
type(covid_df)

pandas.core.frame.DataFrame

<img src="series.png" width=400 height=400 />

#### TODO:

check the data type of the column "new_cases"


In [22]:
covid_df.new_cases.dtypes

dtype('float64')

In [23]:
covid_df.date.dtypes #Object

dtype('O')

Like arrays, you can retrieve a specific value with a series using the indexing notation `[]`.

#### TODO:

Access the last element inside the 'new_cases' column

-1 will not work in here

In [24]:
covid_df.shape

(195, 4)

In [25]:
covid_df.new_cases[covid_df.shape[0]-1]

1326.0

Pandas also provides the `.at` method to retrieve the element at a specific row & column directly.

#### TODO:

Use the .at[] to access the last element inside the 'new_cases' column

In [26]:
covid_df.at[194, 'new_cases']

1326.0

In [27]:
covid_df.at[194, 'new_deaths']

6.0

Instead of using the indexing notation `[]`, Pandas also allows accessing columns as properties of the dataframe using the `.` notation. However, this method only works for columns whose names do not contain spaces or special characters.

#### TODO:

Access the 'new_cases' column using the . notation

In [28]:
covid_df.new_cases,  covid_df['new_cases']

(0        14.0
 1        62.0
 2        53.0
 3        97.0
 4        93.0
         ...  
 190    1444.0
 191    1365.0
 192     996.0
 193     975.0
 194    1326.0
 Name: new_cases, Length: 195, dtype: float64,
 0        14.0
 1        62.0
 2        53.0
 3        97.0
 4        93.0
         ...  
 190    1444.0
 191    1365.0
 192     996.0
 193     975.0
 194    1326.0
 Name: new_cases, Length: 195, dtype: float64)

Further, you can also pass a list of columns within the indexing notation `[]` to access a subset of the data frame with just the given columns.

In [29]:
covid_df

date  new_cases  new_deaths  new_tests
0    2020-02-22       14.0         0.0        NaN
1    2020-02-23       62.0         2.0        NaN
2    2020-02-24       53.0         0.0        NaN
3    2020-02-25       97.0         4.0        NaN
4    2020-02-26       93.0         5.0        NaN
..          ...        ...         ...        ...
190  2020-08-30     1444.0         1.0    53541.0
191  2020-08-31     1365.0         4.0    42583.0
192  2020-09-01      996.0         6.0    54395.0
193  2020-09-02      975.0         8.0        NaN
194  2020-09-03     1326.0         6.0        NaN

[195 rows x 4 columns]

In [30]:
covid_df[['date', 'new_deaths']]

date  new_deaths
0    2020-02-22         0.0
1    2020-02-23         2.0
2    2020-02-24         0.0
3    2020-02-25         4.0
4    2020-02-26         5.0
..          ...         ...
190  2020-08-30         1.0
191  2020-08-31         4.0
192  2020-09-01         6.0
193  2020-09-02         8.0
194  2020-09-03         6.0

[195 rows x 2 columns]

#### TODO:

Access both 'date' and 'new_deaths' columns using subscripts []

In [31]:
cases_df = covid_df[['date', 'new_deaths']]

In [32]:
cases_df #View of original dataframe

date  new_deaths
0    2020-02-22         0.0
1    2020-02-23         2.0
2    2020-02-24         0.0
3    2020-02-25         4.0
4    2020-02-26         5.0
..          ...         ...
190  2020-08-30         1.0
191  2020-08-31         4.0
192  2020-09-01         6.0
193  2020-09-02         8.0
194  2020-09-03         6.0

[195 rows x 2 columns]

The new data frame `cases_df` is simply a "view" of the original data frame `covid_df`. Both point to the same data in the computer's memory. Changing any values inside one of them will also change the respective values in the other. Sharing data between data frames makes data manipulation in Pandas blazing fast. You needn't worry about the overhead of copying thousands or millions of rows every time you want to create a new data frame by operating on an existing one.

Sometimes you might need a full copy of the data frame, in which case you can use the `copy` method.

#### TODO:

Create a copy from covid_df using the .copy() function. Call it covid_df_copy

In [33]:
covid_df_copy = covid_df.copy()

The data within `covid_df_copy` is completely separate from `covid_df`, and changing values inside one of them will not affect the other.

The `.loc` property is used to access a group of rows and columns by label(s) or a boolean array.

.loc[] is primarily label based, but may also be used with a boolean array.


In [34]:
covid_df_copy #Copy of original dataframe 

date  new_cases  new_deaths  new_tests
0    2020-02-22       14.0         0.0        NaN
1    2020-02-23       62.0         2.0        NaN
2    2020-02-24       53.0         0.0        NaN
3    2020-02-25       97.0         4.0        NaN
4    2020-02-26       93.0         5.0        NaN
..          ...        ...         ...        ...
190  2020-08-30     1444.0         1.0    53541.0
191  2020-08-31     1365.0         4.0    42583.0
192  2020-09-01      996.0         6.0    54395.0
193  2020-09-02      975.0         8.0        NaN
194  2020-09-03     1326.0         6.0        NaN

[195 rows x 4 columns]

In [35]:
covid_df

date  new_cases  new_deaths  new_tests
0    2020-02-22       14.0         0.0        NaN
1    2020-02-23       62.0         2.0        NaN
2    2020-02-24       53.0         0.0        NaN
3    2020-02-25       97.0         4.0        NaN
4    2020-02-26       93.0         5.0        NaN
..          ...        ...         ...        ...
190  2020-08-30     1444.0         1.0    53541.0
191  2020-08-31     1365.0         4.0    42583.0
192  2020-09-01      996.0         6.0    54395.0
193  2020-09-02      975.0         8.0        NaN
194  2020-09-03     1326.0         6.0        NaN

[195 rows x 4 columns]

#### TODO:

Access the first location in covid_df using .loc[]

In [36]:
covid_df.loc[0] #gives the first rows values

date          2020-02-22
new_cases           14.0
new_deaths           0.0
new_tests            NaN
Name: 0, dtype: object

In [37]:
covid_df.loc[0]['new_deaths']

0.0

Each retrieved row is also a `Series` object.

#### TODO:

Check the type of the row retrieved using .loc[]

In [38]:
type(covid_df.loc[0]) #Series object

pandas.core.series.Series

`.loc` accepts the same slice notation that Python lists do for both row and columns. Slice notation being start:stop:step
> **Warning**
Note that contrary to usual python slices, **both the start and the stop are included**

#### TODO:

Retrieve the rows starting at 180 and ending at 194 with step size 2 using .loc[]

In [39]:
covid_df.loc[180:194:2] #[start:stop:step size]

date  new_cases  new_deaths  new_tests
180  2020-08-20      642.0         7.0    49662.0
182  2020-08-22      947.0         9.0    46613.0
184  2020-08-24     1209.0         7.0    33358.0
186  2020-08-26      876.0         4.0    58054.0
188  2020-08-28     1409.0         5.0    65135.0
190  2020-08-30     1444.0         1.0    53541.0
192  2020-09-01      996.0         6.0    54395.0
194  2020-09-03     1326.0         6.0        NaN

#### TODO:

Retrieve rows 2 and 190 only from covid_df using .loc[]

In [40]:
covid_df.loc[[2,194]] #Inner brackets is your list, must pass it as a list to location funtion instead of [start:stop:step size]

date  new_cases  new_deaths  new_tests
2    2020-02-24       53.0         0.0        NaN
194  2020-09-03     1326.0         6.0        NaN

**Selecting rows and columns simultaneously** <br>
You have to pass parameters for both row and column inside the  `loc` indexers to select rows and columns simultaneously. The rows and column values may be scalar values, lists, slice objects or boolean.

#### TODO:

Select rows 180 till 194 from the 'new_cases' column

In [41]:
covid_df.loc[180:194, 'new_cases']

180     642.0
181     840.0
182     947.0
183    1071.0
184    1209.0
185     953.0
186     876.0
187    1366.0
188    1409.0
189    1460.0
190    1444.0
191    1365.0
192     996.0
193     975.0
194    1326.0
Name: new_cases, dtype: float64

In [42]:
covid_df.new_cases.loc[180:194]

180     642.0
181     840.0
182     947.0
183    1071.0
184    1209.0
185     953.0
186     876.0
187    1366.0
188    1409.0
189    1460.0
190    1444.0
191    1365.0
192     996.0
193     975.0
194    1326.0
Name: new_cases, dtype: float64

#### TODO:

Select rows 180 till 194 from both 'new_cases' and 'new_deaths' columns

In [43]:
covid_df.loc[180:194, ['new_cases', 'new_deaths']] #Multiple columns need to be passed as list

new_cases  new_deaths
180      642.0         7.0
181      840.0         6.0
182      947.0         9.0
183     1071.0         3.0
184     1209.0         7.0
185      953.0         4.0
186      876.0         4.0
187     1366.0        13.0
188     1409.0         5.0
189     1460.0         9.0
190     1444.0         1.0
191     1365.0         4.0
192      996.0         6.0
193      975.0         8.0
194     1326.0         6.0

In [44]:
covid_df[['new_cases', 'new_deaths']].loc[180:194]

new_cases  new_deaths
180      642.0         7.0
181      840.0         6.0
182      947.0         9.0
183     1071.0         3.0
184     1209.0         7.0
185      953.0         4.0
186      876.0         4.0
187     1366.0        13.0
188     1409.0         5.0
189     1460.0         9.0
190     1444.0         1.0
191     1365.0         4.0
192      996.0         6.0
193      975.0         8.0
194     1326.0         6.0

We can use the `.head` and `.tail` methods to view the first or last few rows of data.

In [45]:
covid_df[['new_cases', 'new_deaths']][180:194]

new_cases  new_deaths
180      642.0         7.0
181      840.0         6.0
182      947.0         9.0
183     1071.0         3.0
184     1209.0         7.0
185      953.0         4.0
186      876.0         4.0
187     1366.0        13.0
188     1409.0         5.0
189     1460.0         9.0
190     1444.0         1.0
191     1365.0         4.0
192      996.0         6.0
193      975.0         8.0

#### TODO:

Print the first 5 rows using .head()

In [46]:
covid_df.head()

date  new_cases  new_deaths  new_tests
0  2020-02-22       14.0         0.0        NaN
1  2020-02-23       62.0         2.0        NaN
2  2020-02-24       53.0         0.0        NaN
3  2020-02-25       97.0         4.0        NaN
4  2020-02-26       93.0         5.0        NaN

#### TODO:

Print the last 4 rows using .tail()

In [47]:
covid_df.tail(4)

date  new_cases  new_deaths  new_tests
191  2020-08-31     1365.0         4.0    42583.0
192  2020-09-01      996.0         6.0    54395.0
193  2020-09-02      975.0         8.0        NaN
194  2020-09-03     1326.0         6.0        NaN

Notice above that while the first few values in the `new_cases` and `new_deaths` columns are `0`, the corresponding values within the `new_tests` column are `NaN`. That is because the CSV file does not contain any data for the `new_tests` column for specific dates (you can verify this by looking into the file). These values may be missing or unknown.

The distinction between `0` and `NaN` is subtle but important. In this dataset, it represents that daily test numbers were not reported on specific dates. Italy started reporting daily tests on Apr 19, 2020. 93,5310 tests had already been conducted before Apr 19. 

We can find the first index that doesn't contain a `NaN` value using a column's `first_valid_index` method.

#### TODO:

Use the .first_valid_index() to check which first index that does not contain NaN values

In [48]:
covid_df.new_tests.first_valid_index() #Must specify row (new_tests)

58

We can use the `.sample` method to retrieve a random sample of rows from the data frame.

#### TODO

Use .sample() to get 10 random row sample from covid_df

In [49]:
covid_df.sample(10) #10 random samples

date  new_cases  new_deaths  new_tests
67   2020-04-29     2091.0       382.0    38589.0
43   2020-04-05     4805.0       681.0        NaN
13   2020-03-06      769.0        41.0        NaN
170  2020-08-10      463.0         2.0    10904.0
151  2020-07-22      128.0        15.0    29288.0
106  2020-06-07      270.0        72.0    27894.0
108  2020-06-09      280.0        65.0    32200.0
22   2020-03-15     3497.0       173.0        NaN
153  2020-07-24      306.0        10.0    28970.0
188  2020-08-28     1409.0         5.0    65135.0

Notice that even though we have taken a random sample, each row's original index is preserved - this is a useful property of data frames.



Here's a summary of the functions & methods we looked at in this section:

- `covid_df['new_cases']` - Retrieving columns as a `Series` using the column name
- `new_cases[243]` - Retrieving values from a `Series` using an index
- `covid_df.at[243, 'new_cases']` - Retrieving a single value from a data frame
- `covid_df.copy()` - Creating a deep copy of a data frame
- `covid_df.loc[243]` - Retrieving a row or range of rows of data from the data frame
- `head`, `tail`, and `sample` - Retrieving multiple rows of data from the data frame
- `covid_df.new_tests.first_valid_index` - Finding the first non-empty index in a series

## Analyzing data from data frames

Let's try to answer some questions about our data.

**Q: What are the total number of reported cases and deaths related to Covid-19 in Italy?**

Similar to Numpy arrays, a Pandas series supports the `sum` method to answer these questions.

In [50]:
covid_df 

date  new_cases  new_deaths  new_tests
0    2020-02-22       14.0         0.0        NaN
1    2020-02-23       62.0         2.0        NaN
2    2020-02-24       53.0         0.0        NaN
3    2020-02-25       97.0         4.0        NaN
4    2020-02-26       93.0         5.0        NaN
..          ...        ...         ...        ...
190  2020-08-30     1444.0         1.0    53541.0
191  2020-08-31     1365.0         4.0    42583.0
192  2020-09-01      996.0         6.0    54395.0
193  2020-09-02      975.0         8.0        NaN
194  2020-09-03     1326.0         6.0        NaN

[195 rows x 4 columns]

#### TODO:

use the .sum() function to find the total cases and total death in covid_df

In [51]:
covid_df.new_cases.sum() #New cases

271512.0

In [52]:
covid_df.new_deaths.sum() #New deaths

35497.0

#### TODO
**Q: What is the overall death rate (ratio of reported deaths to reported cases)?**

In [53]:
death_rate = covid_df.new_deaths.sum() / covid_df.new_cases.sum()

In [54]:
death_rate

0.130738236247385

#### TODO:
**Q: What is the overall number of tests conducted? A total of 935310 tests were conducted before daily test numbers were reported.**


In [55]:
initial_tests = 935310

In [56]:
total_tests = initial_tests + covid_df.new_tests.sum()

In [57]:
total_tests #Pandas ignores rows with NaNs

5214766.0

#### TODO:
**Q: What fraction of tests returned a positive result?**

In [58]:
covid_df.new_cases.sum() / total_tests # 5 percent positivity rate

0.05206599874280073

## Querying and sorting rows

Let's say we only want to look at the days which had more than 1000 reported cases. We can use a boolean expression to check which rows satisfy this criterion.

In [59]:
import numpy as np
arr1 = np.array([[1, 2, 3], [3, 4, 5]])
arr2 = arr1 > 1
arr2

array([[False,  True,  True],
       [ True,  True,  True]])

pandas filtering

In [60]:
covid_df.new_cases

0        14.0
1        62.0
2        53.0
3        97.0
4        93.0
        ...  
190    1444.0
191    1365.0
192     996.0
193     975.0
194    1326.0
Name: new_cases, Length: 195, dtype: float64

In [61]:
high_new_cases = covid_df.new_cases > 1000
high_new_cases # Boolean variable

0      False
1      False
2      False
3      False
4      False
       ...  
190     True
191     True
192    False
193    False
194     True
Name: new_cases, Length: 195, dtype: bool

The boolean expression returns a series containing `True` and `False` boolean values. You can use this series to select a subset of rows from the original dataframe, corresponding to the `True` values in the series.

In [62]:
covid_df.new_cases.loc[high_new_cases] #Only returns where high_new_cases is TRUE

15     1247.0
16     1492.0
17     1797.0
19     2313.0
20     2651.0
        ...  
188    1409.0
189    1460.0
190    1444.0
191    1365.0
194    1326.0
Name: new_cases, Length: 72, dtype: float64

We can write this succinctly on a single line by passing the boolean expression as an index to the data frame.

#### TODO:

Print the entries where the 'new_cases' are greater than 1000 and the 'new_deaths' are greater than 100

In [63]:
covid_df.loc[(covid_df.new_cases > 1000) & (covid_df.new_deaths > 100)]

date  new_cases  new_deaths  new_tests
16  2020-03-09     1492.0       133.0        NaN
19  2020-03-12     2313.0       196.0        NaN
20  2020-03-13     2651.0       189.0        NaN
21  2020-03-14     2547.0       252.0        NaN
22  2020-03-15     3497.0       173.0        NaN
..         ...        ...         ...        ...
75  2020-05-07     1444.0       369.0    13665.0
76  2020-05-08     1401.0       274.0    45428.0
77  2020-05-09     1327.0       243.0    36091.0
78  2020-05-10     1083.0       194.0    31384.0
81  2020-05-13     1402.0       172.0    37049.0

[62 rows x 4 columns]

#### TODO:

Print the entries where the 'new_cases' are greater than 1000 and the 'new_deaths' are greater than 100 and the 'new_tests' are greater than 1000

In [64]:
high_case_df = covid_df.loc[(covid_df.new_cases > 1000) & (covid_df.new_deaths > 100) & (covid_df.new_tests > 1000)]

In [65]:
high_case_df

date  new_cases  new_deaths  new_tests
58  2020-04-20     3047.0       433.0     7841.0
59  2020-04-21     2256.0       454.0    28095.0
60  2020-04-22     2729.0       534.0    44248.0
61  2020-04-23     3370.0       437.0    37083.0
62  2020-04-24     2646.0       464.0    95273.0
63  2020-04-25     3021.0       420.0    38676.0
64  2020-04-26     2357.0       415.0    24113.0
65  2020-04-27     2324.0       260.0    26678.0
66  2020-04-28     1739.0       333.0    37554.0
67  2020-04-29     2091.0       382.0    38589.0
68  2020-04-30     2086.0       323.0    41441.0
69  2020-05-01     1872.0       285.0    43732.0
70  2020-05-02     1965.0       269.0    31231.0
71  2020-05-03     1900.0       474.0    27047.0
72  2020-05-04     1389.0       174.0    22999.0
73  2020-05-05     1221.0       195.0    32211.0
74  2020-05-06     1075.0       236.0    37771.0
75  2020-05-07     1444.0       369.0    13665.0
76  2020-05-08     1401.0       274.0    45428.0
77  2020-05-09     1327.0       243.0    36091.0
78  2020-05-10     1083.0       194.0    31384.0
81  2020-05-13     1402.0       172.0    37049.0

#### TODO:

Using positive_rate from above, print all the rows that has higher rate than the positive_rate

In [66]:
positive_rate = covid_df.new_cases.sum() / covid_df.new_tests.sum()

In [67]:
positive_rate

0.06344544727180278

In [68]:
covid_df.loc[covid_df.new_cases / covid_df.new_tests > positive_rate] #Worst days in italys history for Covid

date  new_cases  new_deaths  new_tests
58  2020-04-20     3047.0       433.0     7841.0
59  2020-04-21     2256.0       454.0    28095.0
61  2020-04-23     3370.0       437.0    37083.0
63  2020-04-25     3021.0       420.0    38676.0
64  2020-04-26     2357.0       415.0    24113.0
65  2020-04-27     2324.0       260.0    26678.0
71  2020-05-03     1900.0       474.0    27047.0
75  2020-05-07     1444.0       369.0    13665.0

The result of performing an operation on two columns is a new series.

We can use this series to add a new column to the data frame.

#### TODO:

Create a new column and call it 'positive_rate'. This column is the 'new_cases' divided by the 'new_tests'

In [69]:
covid_df['positive_rate'] = covid_df.new_cases / covid_df.new_tests

In [70]:
covid_df # Alot of nans, not the best column to have

date  new_cases  new_deaths  new_tests  positive_rate
0    2020-02-22       14.0         0.0        NaN            NaN
1    2020-02-23       62.0         2.0        NaN            NaN
2    2020-02-24       53.0         0.0        NaN            NaN
3    2020-02-25       97.0         4.0        NaN            NaN
4    2020-02-26       93.0         5.0        NaN            NaN
..          ...        ...         ...        ...            ...
190  2020-08-30     1444.0         1.0    53541.0       0.026970
191  2020-08-31     1365.0         4.0    42583.0       0.032055
192  2020-09-01      996.0         6.0    54395.0       0.018311
193  2020-09-02      975.0         8.0        NaN            NaN
194  2020-09-03     1326.0         6.0        NaN            NaN

[195 rows x 5 columns]

However, keep in mind that sometimes it takes a few days to get the results for a test, so we can't compare the number of new cases with the number of tests conducted on the same day. Any inference based on this `positive_rate` column is likely to be incorrect. It's essential to watch out for such subtle relationships that are often not conveyed within the CSV file and require some external context. It's always a good idea to read through the documentation provided with the dataset or ask for more information.

For now, let's remove the `positive_rate` column using the `drop` method.

#### TODO:

Remove the 'positive_rate' column using .dop()

In [71]:
covid_df.drop(columns=['positive_rate'], inplace = True) #Removes column

In [72]:
covid_df #No positivity rate

date  new_cases  new_deaths  new_tests
0    2020-02-22       14.0         0.0        NaN
1    2020-02-23       62.0         2.0        NaN
2    2020-02-24       53.0         0.0        NaN
3    2020-02-25       97.0         4.0        NaN
4    2020-02-26       93.0         5.0        NaN
..          ...        ...         ...        ...
190  2020-08-30     1444.0         1.0    53541.0
191  2020-08-31     1365.0         4.0    42583.0
192  2020-09-01      996.0         6.0    54395.0
193  2020-09-02      975.0         8.0        NaN
194  2020-09-03     1326.0         6.0        NaN

[195 rows x 4 columns]

Can you figure the purpose of the `inplace` argument?

### Sorting rows using column values

The rows can also be sorted by a specific column using `.sort_values`. Let's sort to identify the days with the highest number of cases, then chain it with the `head` method to list just the first ten results.

#### TODO:

Using sort_values(), sort the values based on the 'new_cases' and print the first 10 rows using .head()

inside .sort_values(), make sure to set the ascending parameter to False 

In [73]:
covid_df.sort_values('new_cases', ascending=False).head() #ascending = false starts at the highest and goes down, asc = true starts at lowestand goes up.

date  new_cases  new_deaths  new_tests
29  2020-03-22     6557.0       795.0        NaN
34  2020-03-27     6153.0       660.0        NaN
28  2020-03-21     5986.0       625.0        NaN
36  2020-03-29     5974.0       887.0        NaN
35  2020-03-28     5959.0       971.0        NaN

It looks like the last two weeks of March had the highest number of daily cases. Let's compare this to the days where the highest number of deaths were recorded.

#### TODO:

Using sort_values(), sort the values based on the 'new_deaths' and print the first 10 rows using .head()

In [74]:
covid_df.sort_values('new_deaths', ascending=False).head() #days with the highest deaths

date  new_cases  new_deaths  new_tests
35  2020-03-28     5959.0       971.0        NaN
36  2020-03-29     5974.0       887.0        NaN
39  2020-04-01     4053.0       839.0        NaN
38  2020-03-31     4050.0       810.0        NaN
29  2020-03-22     6557.0       795.0        NaN

It appears that daily deaths hit a peak just about a week after the peak in daily new cases.

Let's also look at the days with the least number of cases. We might expect to see the first few days of the year on this list.

#### TODO:

sort_values() of 'new_cases' and print the first 10 rows. You don't need to set the parameter ascending

In [75]:
covid_df.sort_values('new_cases').head()

date  new_cases  new_deaths  new_tests
119  2020-06-20     -148.0        47.0    29875.0
0    2020-02-22       14.0         0.0        NaN
2    2020-02-24       53.0         0.0        NaN
1    2020-02-23       62.0         2.0        NaN
5    2020-02-27       78.0         1.0        NaN

It seems like the count of new cases on Jun 20, 2020, was `-148`, a negative number! Not something we might have expected, but that's the nature of real-world data. It could be a data entry error, or the government may have issued a correction to account for miscounting in the past. Can you dig through news articles online and figure out why the number was negative?